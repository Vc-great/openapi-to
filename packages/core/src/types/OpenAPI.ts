import type { OpenAPIV2, OpenAPIV3, OpenAPIV3_1 } from 'openapi-types'

export type HttpMethod = `${OpenAPIV3.HttpMethods}`

export type PathGroup = {
  path: string
  method: HttpMethod
  tag: string
}

export type PathGroupByTag = {
  [k in string]: Array<PathGroup>
}

export type OpenAPIDocument = OpenAPIV3.Document | OpenAPIV3_1.Document

export type OpenAPIAllDocument = OpenAPIV2.Document | OpenAPIV3.Document | OpenAPIV3_1.Document
