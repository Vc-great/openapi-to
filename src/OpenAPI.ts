import { OpenAPIV3 } from "openapi-types";
import type { ApiData, ApiNameCache, Config, Parameter } from "./types";
import { Path } from "./Path";
import { Query } from "./Query";
import { RequestBody } from "./RequestBody";
import { Response } from "./Response";
import { Schema } from "./Schema";
import _ from "lodash";

export class OpenAPI {
  public apiItem: ApiData;
  public path: Path;
  public query: Query;
  public requestBody: RequestBody;
  public response: Response;
  public schemas: Schema;
  apiNameCache: ApiNameCache;
  pendingRefCache: Set<string>;

  constructor(
    public openApi3SourceData: OpenAPIV3.Document,
    public config: Config
  ) {
    this.openApi3SourceData = openApi3SourceData;
    this.config = config;
    this.apiNameCache = new Map();
    //缓存$ref
    this.pendingRefCache = new Set();

    this.path = new Path(this);
    this.query = new Query(this);
    this.requestBody = new RequestBody(this);
    this.response = new Response(this);
    this.schemas = new Schema(this);
  }

  get queryRequestName() {
    return `${_.upperFirst(this.apiItem.requestName)}QueryRequest`;
  }

  get pathRequestName() {
    return `${_.upperFirst(this.apiItem.requestName)}PathRequest`;
  }

  get bodyRequestName() {
    return `${_.upperFirst(this.apiItem.requestName)}BodyRequest`;
  }

  get responseName() {
    return `${_.upperFirst(this.apiItem.requestName)}Response`;
  }

  get lowerFirstQueryRequestName() {
    return _.lowerFirst(this.queryRequestName);
  }

  get lowerFirstPathRequestName() {
    return _.lowerFirst(this.pathRequestName);
  }

  get lowerFirstBodyRequestName() {
    return _.lowerFirst(this.bodyRequestName);
  }

  get lowFirstResponseName() {
    return _.lowerFirst(this.responseName);
  }

  traverseParameters(
    parameters: (OpenAPIV3.ReferenceObject | OpenAPIV3.ParameterObject)[],
    {
      has$ref,
      $refInItemSchema,
      other,
    }: {
      has$ref?: Parameter.has$ref;
      $refInItemSchema?: Parameter.$refInItemSchema;
      other: Parameter.other;
    }
  ) {
    if (!parameters) {
      return "";
    }
    const itemTypeMap = (
      parameters: (OpenAPIV3.ReferenceObject | OpenAPIV3.ParameterObject)[]
    ): (string | undefined)[] => {
      return _.map(parameters, (item) => {
        if ("$ref" in item) {
          //todo
          return has$ref && has$ref(item.$ref);
        }

        if (item.schema && "$ref" in item.schema) {
          //todo
          return (
            $refInItemSchema &&
            $refInItemSchema({ item, $ref: item.schema.$ref })
          );
        }

        return other && other({ item, schema: item.schema });
      });
    };
    const joinItem = (itemTypeMap: string[]) => _.join(itemTypeMap, "\n");
    const filterItem = (itemTypeMap: (string | undefined)[]) =>
      _.filter(itemTypeMap, (x: string | undefined) => x);
    return _.flow(itemTypeMap, filterItem, joinItem)(parameters);
  }
}
