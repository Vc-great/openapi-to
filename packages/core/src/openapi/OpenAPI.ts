import _ from "lodash";
import { isRef } from "oas/types";
import { findSchemaDefinition } from "oas/utils";

import { Component } from "./Component.ts";
import { Parameter } from "./Parameter.ts";
import { RequestBody } from "./RequestBody.ts";
import { Response } from "./Response.ts";
import { Schema } from "./Schema.ts";

import type Oas from "oas";
import type { Operation } from "oas/operation";
import type OasTypes from "oas/types";
import type { OpenAPIV3 } from "openapi-types";
import type { OpenAPIV3_1 } from "openapi-types";
import type { HttpMethod, PathGroup, PathGroupByTag } from "../types.js";

const enum WriteModel {
  await = "await",
  succeed = "succeed",
}

type RefCacheValue = {
  schema?: OasTypes.SchemaObject;
  type?: "";
  isWriteIndex?: boolean;
  //等待写入 写入完成
  writeModel?: `${WriteModel}`;
};

export class OpenAPI {
  public parameter: Parameter | undefined;
  public requestBody: RequestBody | undefined;
  public response: Response | undefined;
  public component: Component = new Component(this);
  public schema: Schema | undefined;
  public operation: Operation | undefined;
  public currentTagMetadata: OpenAPIV3.TagObject | undefined;
  public refCache: Map<string, RefCacheValue | null> = new Map<
    string,
    RefCacheValue | null
  >();
  constructor(
    public config: object,
    public oas: Oas,
  ) {
    this.config = config;
    this.oas = oas;
  }

  get currentTagName() {
    return _.camelCase(this.currentTagMetadata && this.currentTagMetadata.name);
  }

  get upperFirstCurrentTagName(): string {
    return _.upperFirst(this.currentTagName);
  }

  //file name and class name
  get operationName(): string {
    return "";
  }
  get methodName(): string {
    return this.generateMethodName();
  }

  //crud的requestName
  get requestName(): string {
    return this.methodName;
  }
  get upperFirstRequestName(): string {
    return _.upperFirst(this.methodName);
  }

  get queryRequestName(): string {
    return `${_.camelCase(this.methodName)}QueryParams`;
  }

  get upperFirstQueryRequestName(): string {
    return `${_.upperFirst(_.camelCase(this.methodName))}QueryParams`;
  }

  get pathRequestName(): string {
    return `${_.camelCase(this.methodName)}PathParams`;
  }

  get upperFirstPathRequestName(): string {
    return `${_.upperFirst(_.camelCase(this.methodName))}PathParams`;
  }

  get bodyRequestName(): string {
    return `${_.camelCase(this.methodName)}BodyParams`;
  }

  get upperFirstBodyRequestName(): string {
    return `${_.upperFirst(_.camelCase(this.methodName))}BodyParams`;
  }

  get responseName(): string {
    return `${_.camelCase(this.methodName)}Response`;
  }
  get upperFirstResponseName(): string {
    return `${_.upperFirst(_.camelCase(this.methodName))}Response`;
  }

  get queryRequestNameForLowerFirst(): string {
    return _.lowerFirst(this.queryRequestName);
  }

  get pathRequestNameForLowerFirst(): string {
    return _.lowerFirst(this.pathRequestName);
  }

  get bodyRequestNameForLowerFirst(): string {
    return _.lowerFirst(this.bodyRequestName);
  }

  get responseNameForLowFirst(): string {
    return _.lowerFirst(this.responseName);
  }

  get pathGroupByTag(): PathGroupByTag {
    return _.chain(this.oas.getPaths())
      .map((item) => _.map(item))
      .flatten()
      .map(({ path, method, schema }) => {
        return _.map(schema.tags, (tag) => {
          return {
            path,
            method,
            tag,
          };
        });
      })
      .flatten()
      .groupBy("tag")
      .value();
  }

  setCurrentOperation(
    path: string,
    method: HttpMethod,
    tagName: string,
  ): Operation {
    this.operation = this.oas.operation(path, method);
    //
    this.currentTagMetadata = this.operation
      .getTags()
      .find((tag) => tag.name === tagName);
    //
    this.parameter = new Parameter(this.operation);
    this.requestBody = new RequestBody(this.operation);
    this.response = new Response(this.operation);
    this.schema = new Schema(this.operation, this);
    return this.operation;
  }

  private generateMethodName(): string {
    if (!this.currentTagMetadata || !this.operation) {
      //todo error log
      return "";
    }
    const tag = this.currentTagMetadata.name;
    const pathGroup = this.pathGroupByTag;
    const crudPath = this.getCrudRequestPath(pathGroup[tag]);
    const methodUpperCase = this.operation.method.toUpperCase();

    //todo
    const currentPath = this.operation.path;
    const lasePathParams = this.getPathLastParams(currentPath);
    const name = new Map([
      ["GET", "findAll"],
      ["POST", "create"],
      ["PUT", "update"],
      ["DELETE", "remove"],
      ["PATCH", "patch"],
      ["DETAIL", "find"],
    ]);

    const isMatch = currentPath === crudPath && name.has(methodUpperCase);

    if (isMatch) {
      return name.get(methodUpperCase) || "";
    }

    //单个删除
    if (methodUpperCase === "DELETE" && lasePathParams) {
      return `delBy${_.upperFirst(_.camelCase(lasePathParams))}`;
    }

    const paths = currentPath.split("/").filter((x) => x);
    const hasBracket = _.last(paths)?.includes("{") || false;

    const detailPath = currentPath.replace(/\/{([^/]+)}/g, "");

    const isDetail = detailPath === crudPath && methodUpperCase === "GET";

    //detail
    if (hasBracket && isDetail) {
      return name.get("DETAIL") + "By" + lasePathParams;
    }
    //其他带括号
    if (hasBracket) {
      const popItem = _.last(paths) || "";

      return (
        _.camelCase(popItem.slice(1, popItem.length - 1)) +
        _.upperFirst(this.operation.method)
      );
    }

    return (
      _.camelCase(paths[paths.length - 1]) + _.upperFirst(this.operation.method)
    );
  }

  //最短路径为crud路径
  public getCrudRequestPath(list: Array<PathGroup> | undefined): string {
    if (list === undefined) {
      return "";
    }
    const obj = {
      path: "",
      length: Number.MAX_SAFE_INTEGER,
    };
    list.forEach((item) => {
      const paths = item.path.split("/").filter((x) => x);
      if (paths.length < obj.length) {
        obj.length = paths.length;
        obj.path = item.path;
      }
    });
    return obj.path;
  }

  //获取路径中最后一个参数
  getPathLastParams(path: string): string {
    const pathParts = path.split("/");

    const part = _.last(pathParts) || "";
    if (part.startsWith("{") && part.endsWith("}")) {
      return _.upperFirst(_.camelCase(part.slice(1, -1)));
    } else {
      return part;
    }
  }

  /**
   * @param check Data to determine if it contains a ReferenceObject (`$ref` pointer`).
   * @returns If the supplied data has a `$ref` pointer.
   */
  isReference(
    check: unknown,
  ): check is OpenAPIV3.ReferenceObject | OpenAPIV3_1.ReferenceObject {
    return isRef(check);
  }

  resetRefCache(): void {
    this.refCache.clear();
  }

  getRefAlias($ref: string): string {
    //todo ref cache
    this.refCache.set($ref, null);
    const typeName = $ref.replace(/.+\//, "");
    return _.camelCase(typeName);
  }

  hasRefByCache($ref: string): boolean {
    return [...this.refCache.keys()].includes($ref);
  }

  setRefCache(ref: string, config: RefCacheValue): void {
    const value = this.refCache.get(ref);
    this.refCache.set(ref, {
      ...value,
      ...config,
    });
  }

  /**
   * Lookup a reference pointer within an OpenAPI definition and return the schema that it resolves
   * to.
   *
   * @param $ref Reference to look up a schema for.
   */
  findSchemaBy$ref($ref: string): any {
    const schema = findSchemaDefinition(
      $ref,
      this.operation?.api,
    ) as OasTypes.SchemaObject;

    this.refCache.set($ref, {
      schema: schema,
      writeModel: "await",
    });
    return schema;
  }
}
