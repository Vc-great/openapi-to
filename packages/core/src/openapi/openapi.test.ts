import Oas from 'oas'

import petStore from '../../mock/petstore.json'
import { OpenAPI } from './OpenAPI.ts'

describe('openapi get requestName', () => {
  const oas = new Oas(petStore)
  const openapi = new OpenAPI({}, oas)

  test('request crud name :create', () => {
    openapi.setCurrentOperation('/pet', 'post', 'pet')
    const requestName = openapi.requestName
    expect(requestName).toBe('create')
  })

  test('request crud name:findAll', () => {
    openapi.setCurrentOperation('/pet', 'get', 'pet')
    const requestName = openapi.requestName
    expect(requestName).toBe('findAll')
  })

  test('request crud name:remove', () => {
    openapi.setCurrentOperation('/pet', 'delete', 'pet')
    const requestName = openapi.requestName
    expect(requestName).toBe('remove')
  })

  test('request crud name:update', () => {
    openapi.setCurrentOperation('/pet', 'put', 'pet')
    const requestName = openapi.requestName
    expect(requestName).toBe('update')
  })

  test('request crud name:findById', () => {
    openapi.setCurrentOperation('/pet/{id}', 'get', 'pet')
    const requestName = openapi.requestName
    expect(requestName).toBe('findById')
  })

  test('request name:{name}Get', () => {
    openapi.setCurrentOperation('/pet/findByStatus', 'get', 'pet')
    const requestName = openapi.requestName
    expect(requestName).toBe('findByStatusGet')
  })
})

describe('openapi get pathGroupByTag', () => {
  const oas = new Oas(petStore)
  const openapi = new OpenAPI({}, oas)
  test('get pathGroupByTag', () => {
    const pathGroup = openapi.pathGroupByTag
    expect(pathGroup).toMatchSnapshot()
  })
})
