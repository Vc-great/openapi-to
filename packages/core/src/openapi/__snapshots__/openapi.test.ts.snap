// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`openapi component > requestBodyObject 1`] = `
{
  "Pet": {
    "content": {
      "application/json": {
        "schema": {
          "$ref": "#/components/schemas/Pet",
        },
      },
      "application/xml": {
        "schema": {
          "$ref": "#/components/schemas/Pet",
        },
      },
    },
    "description": "Pet object that needs to be added to the store",
    "required": true,
  },
  "UserArray": {
    "content": {
      "application/json": {
        "schema": {
          "items": {
            "$ref": "#/components/schemas/User",
          },
          "type": "array",
        },
      },
    },
    "description": "List of user object",
    "required": true,
  },
  "testDto": {
    "content": {
      "application/json": {
        "schema": {
          "$ref": "#/components/schemas/testDto",
        },
      },
    },
    "description": "resources",
    "required": true,
  },
}
`;

exports[`openapi component > responsesObject 1`] = `null`;

exports[`openapi get pathGroupByTag > get pathGroupByTag 1`] = `
{
  "pet": [
    {
      "method": "post",
      "path": "/pet/test",
      "tag": "pet",
    },
    {
      "method": "put",
      "path": "/pet/test",
      "tag": "pet",
    },
    {
      "method": "delete",
      "path": "/pet/test",
      "tag": "pet",
    },
    {
      "method": "get",
      "path": "/pet/test/{id}",
      "tag": "pet",
    },
    {
      "method": "get",
      "path": "/pet/{id}",
      "tag": "pet",
    },
    {
      "method": "post",
      "path": "/pet/{petId}/uploadImage",
      "tag": "pet",
    },
    {
      "method": "get",
      "path": "/pet",
      "tag": "pet",
    },
    {
      "method": "post",
      "path": "/pet",
      "tag": "pet",
    },
    {
      "method": "put",
      "path": "/pet",
      "tag": "pet",
    },
    {
      "method": "delete",
      "path": "/pet",
      "tag": "pet",
    },
    {
      "method": "get",
      "path": "/pet/findByStatus",
      "tag": "pet",
    },
    {
      "method": "get",
      "path": "/pet/findByTags",
      "tag": "pet",
    },
    {
      "method": "get",
      "path": "/pet/{petId}",
      "tag": "pet",
    },
    {
      "method": "post",
      "path": "/pet/{petId}",
      "tag": "pet",
    },
    {
      "method": "delete",
      "path": "/pet/{petId}",
      "tag": "pet",
    },
  ],
  "store": [
    {
      "method": "post",
      "path": "/store/order",
      "tag": "store",
    },
    {
      "method": "get",
      "path": "/store/order/{orderId}",
      "tag": "store",
    },
    {
      "method": "delete",
      "path": "/store/order/{orderId}",
      "tag": "store",
    },
    {
      "method": "get",
      "path": "/store/inventory",
      "tag": "store",
    },
  ],
  "user": [
    {
      "method": "post",
      "path": "/user/createWithArray",
      "tag": "user",
    },
    {
      "method": "post",
      "path": "/user/createWithList",
      "tag": "user",
    },
    {
      "method": "get",
      "path": "/user/{username}",
      "tag": "user",
    },
    {
      "method": "put",
      "path": "/user/{username}",
      "tag": "user",
    },
    {
      "method": "delete",
      "path": "/user/{username}",
      "tag": "user",
    },
    {
      "method": "get",
      "path": "/user/login",
      "tag": "user",
    },
    {
      "method": "get",
      "path": "/user/logout",
      "tag": "user",
    },
    {
      "method": "post",
      "path": "/user",
      "tag": "user",
    },
  ],
}
`;
